### DISCUSSION TITLE
Integration with crispy forms

### DISCUSSION THREAD

**Discussion by adammakowskidev** [django-commons/django-unicorn#227] [Ideas]
Hi.
Is it possible to integrate unicorn with crispy forms?
I usually use crispy forms in my projects
Has anyone tested it and can share their experiences?

**Comment by adamghill** [comment]
I've looked at `django-crispy-forms` in the past, but have never used it, so I'm not sure. However, it looks you might be able to set custom attributes in the form? There is a `attrs` on the form helper mentioned in https://django-crispy-forms.readthedocs.io/en/latest/form_helper.html#helper-attributes-you-can-set. Can you try adding something like `{ "unicorn:model": "FIELD_NAME"}` there and see if it works?

**Reply by nerdoc** [reply]
I need the same, as I have to do many forms in my project, and have some special needs there. Doing everything "by hand" is too much, so crispy forms are great.
I even created a form that is dynamically generated (in a custom Form) from database data, and displayed in a class inheriting from FormView. The template has to be "dynamic" too, as some widgets blend in/out other widgets groups etc.

@adamghill, your approach with attrs does not help much, as this helper is of the `Form` class, and not per widget.
But you can create form fields dynamically and just add widget attrs to them directly, like 

```python
field.widget.attrs["unicorn:model"] = "<field-name>"
```
This works at least as the widgets get this attribute correctly.

What is still a problem here I did not manage yet: You have to inherit from (any child of) FormView to get a FormView. **AND**, you have to inherit from UnicornView, which is a TemplateView child:

```python
class PreferencesView(PermissionRequiredMixin, FormView, UnicornView):
```

Which seems to bite each other. I get this error:

```
TypeError at /preferences
Type is not JSON serializable: __proxy__

Request Method: | GET
-- | --
http://127.0.0.1:8000/preferences
4.0.1
TypeError
Type is not JSON serializable: __proxy__
/.../.venv/lib/python3.10/site-packages/django_unicorn/serializer.py, line 300, in dumps
```



But, @adammakowskidev, If you don't need dynamic Form data, just static ones, take the [Crispy docs for HTML Attrs of Widgets](https://django-crispy-forms.readthedocs.io/en/latest/layouts.html#layout-objects-attributes) - this should do it for you as well, or define it at declaration time direct in the Form:
```python
class FooForm(Form):
    email = forms.EmailField(widget=forms.EmailWidget(attrs={"unicorn:model":"YourModel"}))
```
Not tested, but should do the trick.

I think there is some help as well in [Crispy AJAX validation](https://django-crispy-forms.readthedocs.io/en/latest/crispy_tag_forms.html#ajax-validation-recipe), in case you need no dynamic forms.

The problem is just the View, and the IMHO Inheritance (Diamond of Death?) problem... @adamghill do you have a hint here? How do I combine a FormView with UnicornView?

**Reply by nerdoc** [reply]
AAAAAAHHH, my bad. Has nothing to do with inheritance, Diamond of Death altogether... It was a `success_url = reverse_lazy("preferences")` in the hybrid Form class which created the __proxy__ / JSON error.

After removing the reverse_lazy, everything works now as intended!
